# zuul 统一异常处理
首先需要提示的是 spring-cloud 版本不一致，处理的方式是千差万别的。  
SendErrorFilter
```java
	public boolean shouldFilter() {
		RequestContext ctx = RequestContext.getCurrentContext();
		// only forward to errorPath if it hasn't been forwarded to already
		return ctx.getThrowable() != null
				&& !ctx.getBoolean(SEND_ERROR_FILTER_RAN, false);
	}
```
而其他版本判断的方式不一样。
所有的错误问题的最终都是被SendErrorFilter类来处理，该类在早期的版本是一个post类型的filter，post类型的filter有一个缺陷就是不能处理post中抛出的异常，需要我们手动去完善，而我目前使用的这个版本(Dalston.SR3)已经修复了这个问题，SendErrorFilter现在是一个error类型的filter，而且只要RequestContext中有异常就会进入到SendErrorFilter中，错误信息也都从exception对象中提取出来

发生异常的PRE前置（其他类型也可以）过滤器
```java
package cn.lovingliu.apigateway.filter;

import com.netflix.zuul.ZuulFilter;
import com.netflix.zuul.context.RequestContext;
import com.netflix.zuul.exception.ZuulException;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang.StringUtils;
import org.springframework.cloud.netflix.zuul.filters.support.FilterConstants;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;

/**
 * @Author：LovingLiu
 * @Description: zuul 前置过滤器
 * @Date：Created in 2019-10-18
 */
@Component
@Slf4j
public class TokenFilter extends ZuulFilter {
    @Override
    public String filterType() {
        return FilterConstants.PRE_TYPE;
    }

    @Override
    public int filterOrder() {
        return FilterConstants.PRE_DECORATION_FILTER_ORDER - 1;
    }

    @Override
    public boolean shouldFilter() {
        return true;
    }

    @Override
    public Object run() throws ZuulException {
        RequestContext requestContext = RequestContext.getCurrentContext();
        HttpServletRequest request = requestContext.getRequest();
        // 这里是从URL参数里获取,也可以从cookie和header中获取
        String token = request.getParameter("token");
        if(StringUtils.isBlank(token)){
            log.error("验证未通过,token参数缺失");
            // 统一异常处理
            throw new RuntimeException("验证未通过,token参数缺失");
            //requestContext.setSendZuulResponse(false);
            //requestContext.setResponseStatusCode(HttpStatus.UNAUTHORIZED.value());// 401
        }
        return null;
    }

}

```
## 一.自定义ErrorController

### 1.创建ExceptionController 实现ErrorController
```java
package cn.lovingliu.apigateway.controller;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.web.servlet.error.ErrorController;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.servlet.http.HttpServletRequest;

/**
 * @Author：LovingLiu
 * @Description:
 * @Date：Created in 2019-10-19
 */
@RestController
public class ExceptionController  implements ErrorController {

    @Value("${error.path}")
    private String errorPath;

    @GetMapping(value = "${error.path}")
    public String error(HttpServletRequest request) {
        Exception e = (Exception) request.getAttribute("javax.servlet.error.exception");
        String message = e.getCause().getMessage();
        return message;
    }

    @Override
    public String getErrorPath() {
        return "";
    }
}
```
### 2.配置文件添加error.path
```xml
error:
  path: /error
```

## 二.禁用zuul默认的异常处理filter，使用自定义的ErrorFilter
### 1.禁用Zuul默认的异常处理filter—SendErrorFilter
```xml
zuul:
    SendErrorFilter:
        error:
            disable: true
```
### 2.自定义ErrorFilter
```java
/**
 * 自定义异常处理Filter 需要禁用Zuul自带的异常处理Filter--SendErrorFilter
 * 
 * @author Administrator
 *
 */
@Component
public class ErrorFilter extends ZuulFilter {
	private static final Log log = LogFactory.getLog(ErrorFilter.class);
	protected static final String SEND_ERROR_FILTER_RAN = "sendErrorFilter.ran";
 
	@Override
	public String filterType() {
		return ERROR_TYPE;
	}
 
	@Override
	public int filterOrder() {
		return SEND_ERROR_FILTER_ORDER;
	}
 
	public boolean shouldFilter() {
		RequestContext ctx = RequestContext.getCurrentContext();
		// only forward to errorPath if it hasn't been forwarded to already
		return ctx.getThrowable() != null && !ctx.getBoolean(SEND_ERROR_FILTER_RAN, false);
	}
 
	public Object run() {
		try {
			RequestContext context = RequestContext.getCurrentContext();
			ZuulException exception = this.findZuulException(context.getThrowable());
			log.error("进入系统异常拦截", exception);
 
			HttpServletResponse response = context.getResponse();
			response.setContentType("application/json; charset=utf8");
			response.setStatus(exception.nStatusCode);
			PrintWriter writer = null;
			try {
				writer = response.getWriter();
				writer.print("{code:" + exception.nStatusCode + ",message:\"" + exception.getMessage() + "\"}");
			} catch (IOException e) {
				e.printStackTrace();
			} finally {
				if (writer != null) {
					writer.close();
				}
			}
 
		} catch (Exception var5) {
			ReflectionUtils.rethrowRuntimeException(var5);
		}
 
		return null;
	}
 
	ZuulException findZuulException(Throwable throwable) {
		if (throwable.getCause() instanceof ZuulRuntimeException) {
			// this was a failure initiated by one of the local filters
			return (ZuulException) throwable.getCause().getCause();
		}
 
		if (throwable.getCause() instanceof ZuulException) {
			// wrapped zuul exception
			return (ZuulException) throwable.getCause();
		}
 
		if (throwable instanceof ZuulException) {
			// exception thrown by zuul lifecycle
			return (ZuulException) throwable;
		}
 
		// fallback, should never get here
		return new ZuulException(throwable, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, null);
	}
}
```
## 3.第三种方式，重写错误信息(不推荐)


# 坑:
同样的参数,同样的接口,GET请求能被ZUUl捕获然后重定向到 /error,但是post能捕获,但不能重定向到/error
